ndk/src/event.rs

```diff
@@ -47,18 +47,30 @@ pub enum Source {
     Any = ffi::AINPUT_SOURCE_ANY,
 }
 
-/// An enum representing the class of an [`InputEvent`] source.
-///
-/// See [the NDK docs](https://developer.android.com/ndk/reference/group/input#anonymous-enum-35)
-#[derive(Debug, Clone, Copy, PartialEq, Eq, TryFromPrimitive, IntoPrimitive)]
-#[repr(u32)]
-enum Class {
-    None = ffi::AINPUT_SOURCE_CLASS_NONE,
-    Button = ffi::AINPUT_SOURCE_CLASS_BUTTON,
-    Pointer = ffi::AINPUT_SOURCE_CLASS_POINTER,
-    Navigation = ffi::AINPUT_SOURCE_CLASS_NAVIGATION,
-    Position = ffi::AINPUT_SOURCE_CLASS_POSITION,
-    Joystick = ffi::AINPUT_SOURCE_CLASS_JOYSTICK,
+impl Source {
+    pub fn class(self) -> SourceClass {
+        let class = u32::from(self) & ffi::AINPUT_SOURCE_CLASS_MASK;
+        SourceClass::from_bits_retain(class)
+    }
+}
+
+bitflags::bitflags! {
```

Comment:
If the bitflags API is exposed directly I wonder if we should add a a compatibility member like `const _ = !0;` so `all()` wouldn't assume these flags won't be extended and would set all bits:

> [Externally defined flags](https://docs.rs/bitflags/latest/bitflags/#externally-defined-flags)
> If you’re generating flags types for an external source, such as a C API, you can define an extra unnamed flag as a  mask of all bits the external source may ever set. Usually this would be all bits (!0):

I guess we could remove the `AINPUT_SOURCE_CLASS_NONE` which is zero - so not actually a flag. Instead `::empty()` and `::is_empty()` can be used from the bitflags api.

Reply:
That's why we haven't used `bitflags` in the `ash` vulkan crate, and even removed the `all()` and `not()` inversion methods a long time ago. That was also because `not()` was implemented as `self ^ all()`, so only known-at-compile-time bits get flipped. Fortunately `bitflags` implements it differently.

Let's stick with the described workaround.

Also yeah, I've been thinking about the `NONE` flag, which now has an alias and is represented upstream.